modules/standard/Path.chpl

```diff
@@ -300,41 +301,205 @@ proc file.realPath(): string throws {
    :rtype: `string`
 */
   proc expandVars(path: string): string{
-    var path_expanded: string = path;
-    var env_arr = [("CHPL_HOME", CHPL_HOME),
-                   ("CHPL_AUX_FILESYS", CHPL_AUX_FILESYS),
-                   ("CHPL_TARGET_PLATFORM", CHPL_TARGET_PLATFORM),
-                   ("CHPL_HOST_PLATFORM", CHPL_HOST_PLATFORM),
-                   ("CHPL_HOST_COMPILER", CHPL_HOST_COMPILER),
-                   ("CHPL_TARGET_COMPILER", CHPL_TARGET_COMPILER),
-                   ("CHPL_TARGET_ARCH", CHPL_TARGET_ARCH),
-                   ("CHPL_LOCALE_MODEL", CHPL_LOCALE_MODEL),
-                   ("CHPL_COMM_SUBSTRATE", CHPL_COMM_SUBSTRATE),
-                   ("CHPL_COMM", CHPL_COMM),
-                   ("CHPL_GASNET_SEGMENT", CHPL_GASNET_SEGMENT),
-                   ("CHPL_TASKS", CHPL_TASKS),
-                   ("CHPL_LAUNCHER", CHPL_LAUNCHER),
-                   ("CHPL_TIMERS", CHPL_TIMERS),
-                   ("CHPL_UNWIND", CHPL_UNWIND),
-                   ("CHPL_MEM", CHPL_MEM),
-                   ("CHPL_MAKE", CHPL_MAKE),
-                   ("CHPL_ATOMICS", CHPL_ATOMICS),
-                   ("CHPL_NETWORK_ATOMICS", CHPL_NETWORK_ATOMICS),
-                   ("CHPL_GMP", CHPL_GMP),
-                   ("CHPL_HWLOC", CHPL_HWLOC),
-                   ("CHPL_REGEXP", CHPL_REGEXP),
-                   //("CHPL_WIDE_POINTERS", CHPL_WIDE_POINTERS),
-                   ("CHPL_LLVM", CHPL_LLVM)
-                  ];
-    for i in 1..23{
-      if(path_expanded.find("$"+env_arr(i)(1))>0){
-        path_expanded = path_expanded.replace("$"+env_arr(i)(1), env_arr(i)(2));
+    var path_p:string = path;
+    var varChars:string = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-_";
+    var res:string = "";
+    var ind:int = 1;
+    var pathlen:int = path_p.length;
+    while(ind <= pathlen){
+        var c:string = path_p(ind);
+        if(c == "$"){
+            if(path_p(ind+1) == "$"){
+                res = res + c;
+                ind += 1;
+            }
+            else if(path_p(ind+1) == "{"){
+                path_p = path_p((ind+2)..);
+                pathlen = path_p.length;
+                ind = path_p.find("}");
```

Comment:
This seems like it assumes the next `}` you find will close the current open one.  Am I wrong in thinking we could have a nested situation with `{ }`, for example `${FOO${BAR}}`?

Reply:
This kind of a nesting in environment variables is not supported in bash. That is why, I excluded handling that case in the code. For reference, https://stackoverflow.com/questions/917260/can-var-parameter-expansion-expressions-be-nested-in-bash .

Reply:
True, but we're not running the function itself in bash, just checking against the contents of the string provided to us.  In the post, it sounds like zsh does support that kind of nesting, so it wouldn't be unheard of for someone to try sending that kind of path to us.  What does Python do when run from bash over that kind of string?

Reply:
I checked what python returns for a nested env variable using the following code:
```
import os
os.environ["ridge"]="bar"
os.environ["foobar"]="wall"
path1 = "${foo${ridge}}"
path2 = os.path.expandvars(path1)
print path2
```
And the output that was returned was ```${foo${ridge}}```, which means that python leaves such variables unchanged.